stdlib/pagerduty/pagerduty.go

```diff
@@ -0,0 +1,211 @@
+	plan.DefaultCost
+}
+
+func (s *DedupProcedureSpec) Kind() plan.ProcedureKind {
+	return DedupKeyKind
+}
+
+func (s *DedupProcedureSpec) Copy() plan.ProcedureSpec {
+	ns := *s
+	return &ns
+}
+
+func newDedupKeyProcedure(qs flux.OperationSpec, pa plan.Administration) (plan.ProcedureSpec, error) {
+	if _, ok := qs.(*DedupKeyOpSpec); !ok {
+		return nil, fmt.Errorf("invalid spec type %T", qs)
+	}
+	return &DedupProcedureSpec{}, nil
+}
+
+type DedupKeyTransformation struct {
+	d     execute.Dataset
+	cache execute.TableBuilderCache
+}
+
+func createDedupKeyTransformation(id execute.DatasetID, mode execute.AccumulationMode, spec plan.ProcedureSpec, a execute.Administration) (execute.Transformation, execute.Dataset, error) {
+	cache := execute.NewTableBuilderCache(a.Allocator())
+	dataset := execute.NewDataset(id, mode, cache)
+	if _, ok := spec.(*DedupProcedureSpec); !ok {
+		return nil, nil, fmt.Errorf("invalid spec type %T", spec)
+	}
+
+	transform := NewDedupKeyTransformation(dataset, cache)
+	return transform, dataset, nil
+}
+
+func NewDedupKeyTransformation(d execute.Dataset, cache execute.TableBuilderCache) *DedupKeyTransformation {
+	return &DedupKeyTransformation{
+		d:     d,
+		cache: cache,
+	}
+}
+
+func (t *DedupKeyTransformation) RetractTable(id execute.DatasetID, key flux.GroupKey) error {
+	return t.d.RetractTable(key)
+}
+
+type kvs struct {
+	k string
+	v string
+}
+
+func (t *DedupKeyTransformation) Process(id execute.DatasetID, tbl flux.Table) error {
+	groupCols := tbl.Key().Cols()
+	groupVals := tbl.Key().Values()
+	keys := make([]kvs, len(groupCols))
+	for i := range groupCols {
+		keys[i].k = groupCols[i].Label
+		v := groupVals[i]
+		var str string
+		switch v.Type().Nature() {
+		case semantic.String:
+			str = v.Str()
+		case semantic.Int:
+			str = strconv.FormatInt(v.Int(), 10)
+		case semantic.UInt:
+			str = strconv.FormatUint(v.UInt(), 10)
+		case semantic.Float:
+			str = strconv.FormatFloat(v.Float(), 'f', -1, 64)
+		case semantic.Bool:
+			str = strconv.FormatBool(v.Bool())
+		case semantic.Time:
+			str = v.Time().String()
+		case semantic.Duration:
+			str = v.Duration().String()
+		default:
+			return fmt.Errorf("cannot convert %v to string", v.Type())
+		}
+
+		keys[i].v = str
+
+	}
+	sort.Slice(keys, func(i, j int) bool {
+		return keys[i].k < keys[j].k
+	})
+
+	sb := strings.Builder{}
+	for i := range keys {
+		sb.WriteString(keys[i].k)
+		sb.WriteRune('\n')
+		sb.WriteString(keys[i].v)
+		sb.WriteRune('\n')
+	}
+
+	dedupKey := sb.String()
+
+	builder, isNew := t.cache.TableBuilder(tbl.Key())
+
+	if isNew {
+		if err := execute.AddTableCols(tbl, builder); err != nil { // adds the other columns to builder
+			return err
+		}
+	}
+
+	err := tbl.Do(func(cr flux.ColReader) error {
+		l := cr.Len()
+		if err := execute.AppendCols(cr, builder); err != nil {
+			return err
+		}
+		colIDX, err := builder.AddCol(flux.ColMeta{
+			Label: dedupKeyColName,
+			Type:  flux.TString,
+		})
```

Comment:
This will add a column to the builder for each col reader in the incoming table which is not what we want. The column should be added once outside the `tbl.Do` loop.

An example of when this would break is when we read data from storage we read it in 1000 row chunks, each chunk becomes its own ColReader. So a table with 3000 rows would have three col readers and would add the column to the builder three times.